core/BTB.hpp

```diff
@@ -0,0 +1,277 @@
+        {
+            if (&rhs != this) {
+                BasicCacheItem::operator=(rhs);
+                valid_ = rhs.valid_;
+                target_ = rhs.target_;
+                is_call_ = rhs.is_call_;
+                branchtype_ = rhs.branchtype_;
+                lhist_counter_ = rhs.lhist_counter_;
+            }
+            return *this;
+        }
+
+        void reset(uint64_t addr)
+        {
+            setValid(true);
+            setAddr(addr);
+            resetLHistCounter_();
+        }
+
+        void setValid(bool v) { valid_ = v; }
+        bool isValid() const { return valid_; }
+
+        void setTarget(uint64_t t) { target_ = t; }
+        uint64_t getTarget() const { return target_; }
+
+        void setBranchType(BranchType branchtype, bool is_call = false)
+        {
+            sparta_assert(!is_call ||
+                          miscutils::isOneOf(branchtype, BranchType::INDIRECT,
+                                                         BranchType::DIRECT,
+                                                         BranchType::RETURN),
+                          "Invalid branch and call combination");
+            branchtype_ = branchtype;
+            is_call_ = is_call;
+        }
+
+        BranchType getBranchType() const { return branchtype_; }
+        bool isCall() const { return is_call_; }
+
+        bool predictDirection() {
+            if (branchtype_ == BranchType::CONDITIONAL)
+            {
+                return lhist_counter_ >= 0;
+            }
+            return true;
+        }
+
+        void updateDirection(bool taken)
+        {
+            if (taken)
+            {
+                lhist_counter_ = std::min(lhist_count_max_, ++lhist_counter_);
+            }
+            else
+            {
+                lhist_counter_ = std::max(lhist_count_min_, --lhist_counter_);
+            }
+        }
+
+
+    private:
+        bool                 valid_ = false;
+        uint64_t             target_;
+        bool                 is_call_ = false;
+        BranchType           branchtype_;
+        int32_t              lhist_counter_ = 0;
+        static constexpr int32_t lhist_count_max_ = 1; // 2 bit saturating counter.
+        static constexpr int32_t lhist_count_min_ = -2;
+
+        void resetLHistCounter_() { lhist_counter_ = 0; }
+
+    };
+
+
+    class BTBAddrDecoder : public sparta::cache::AddrDecoderIF
+    {
+    public:
+        BTBAddrDecoder(uint32_t entries, uint32_t stride, uint32_t num_ways)
+        {
+            index_mask_ = (entries/num_ways)-1;
+            index_shift_ = sparta::utils::floor_log2(stride);
+        }
+        virtual uint64_t calcTag(uint64_t addr) const { return addr; }
+        virtual uint32_t calcIdx(uint64_t addr) const { return (addr >> index_shift_) & index_mask_; }
+        virtual uint64_t calcBlockAddr(uint64_t addr) const { return addr; }
+        virtual uint64_t calcBlockOffset(uint64_t addr) const { return 0;}
```

Comment:
Are these new `virtual` methods or overrides from `AddrDecoderIF`?  If overrides, remove the `virtual` keyword and add `override` to the end.